util/cliutil/earthlydir.go

```diff
@@ -11,96 +11,109 @@ import (
+
+	earthlyDirCreateOnce.Do(func() {
+		if !fileutil.DirExists(earthlyDir) {
+			err := os.MkdirAll(earthlyDir, 0755)
+			if err != nil {
+				earthlyDirCreateErr = errors.Wrapf(err, "unable to create dir %s", earthlyDir)
+				return
+			}
+			if earthlyDirSudoUser != nil {
+				fileutil.EnsureUserOwned(earthlyDir, earthlyDirSudoUser)
+			}
 		}
-	}
+	})
 
-	if u.HomeDir == "" {
-		return "/etc", u, nil
-	}
-	return u.HomeDir, u, nil
+	return earthlyDir, earthlyDirCreateErr
 }
 
-// IsBootstrapped provides a tentatively correct guess about the state of our bootstrapping.
-func IsBootstrapped() bool {
-	homeDir, _, err := DetectHomeDir()
+func getHomeFromSudoUser() (string, *user.User, bool) {
+	sudoUserName, ok := os.LookupEnv("SUDO_USER")
+	if !ok {
+		return "", nil, false
+	}
+	u, err := user.Lookup(sudoUserName)
 	if err != nil {
-		return false
+		return "", nil, false
 	}
-
-	earthlyDir := filepath.Join(homeDir, ".earthly")
-	if !fileutil.DirExists(earthlyDir) {
-		return false
+	if u.HomeDir == "" {
+		return "", nil, false
 	}
-
-	return true
+	return u.HomeDir, u, true
 }
 
-// EnsurePermissions changes the permissions of all earthly files to be owned by the user and their group.
-func EnsurePermissions() error {
-	_, err := GetEarthlyDir()
-	if err != nil {
-		return err
+func getHomeFromHomeEnv() (string, *user.User, bool) {
+	home, ok := os.LookupEnv("HOME")
+	if !ok {
+		return "", nil, false
 	}
+	return home, nil, true
+}
 
-	u, _, err := currentNonSudoUser()
+func getHomeFromUserCurrent() (string, *user.User, bool) {
+	u, err := user.Current()
 	if err != nil {
-		return errors.Wrap(err, "get non-sudo user")
+		return "", nil, false
 	}
-
-	fileutil.EnsureUserOwned(earthlyDir, u)
-	return nil
+	if u.HomeDir == "" {
+		return "", nil, false
+	}
+	return u.HomeDir, u, true
 }
 
-func currentNonSudoUser() (*user.User, bool, error) {
-	if sudoUserName, ok := os.LookupEnv("SUDO_USER"); ok {
-		sudoUser, err := user.Lookup(sudoUserName)
-		if err == nil {
-			return sudoUser, true, nil
+// DetectHomeDir returns the home directory of the current user, together with
+// the user object who owns it. If SUDO_USER is detected, then that user's
+// home directory will be used instead.
+func DetectHomeDir() (string, *user.User) {
+	for _, fn := range []func() (string, *user.User, bool){
+		getHomeFromSudoUser,
+		getHomeFromHomeEnv,
+		getHomeFromUserCurrent,
+	} {
+		home, u, ok := fn()
+		if ok {
+			return home, u
 		}
 	}
+	return "/etc", nil
+}
 
-	u, err := user.Current()
-	if err != nil {
-		return nil, false, err
+// IsBootstrapped provides a tentatively correct guess about the state of our bootstrapping.
+func IsBootstrapped() bool {
+	return fileutil.DirExists(GetEarthlyDir())
+}
+
+// EnsurePermissions changes the permissions of all earthly files to be owned by the user and their group.
+func EnsurePermissions() error {
+	earthlyDir, sudoUser := getEarthlyDirAndUser()
```

Comment:
I *think* that the user is returned even if it's not a `SUDO_USER` case.

I think the logic is ok either way, but perhaps we shouldn't name the variable `sudoUser` just to make it less misleading.

Reply:
that's a good question, I intended on it only ever being the sudo user, but I see we could potentiall return a different user in this function: https://github.com/earthly/earthly/blob/7015379873db0c51b7791b5527961e5f449496ea/util/cliutil/earthlydir.go#L79-L88

I think it's best for us to return nil here as there would be no need to change the permissions for this case.

Reply:
Here's the change for this issue: https://github.com/earthly/earthly/commit/3346674201e92a84c097192bbbc84952fbf9976e